src/main/java/rx/internal/operators/OperatorMaterialize.java

```diff
@@ -29,41 +32,147 @@
  * See <a href="http://msdn.microsoft.com/en-us/library/hh229453.aspx">here</a> for the Microsoft Rx equivalent.
  */
 public final class OperatorMaterialize<T> implements Operator<Notification<T>, T> {
+
     /** Lazy initialization via inner-class holder. */
     private static final class Holder {
         /** A singleton instance. */
         static final OperatorMaterialize<Object> INSTANCE = new OperatorMaterialize<Object>();
     }
+
     /**
      * @return a singleton instance of this stateless operator.
      */
     @SuppressWarnings("unchecked")
     public static <T> OperatorMaterialize<T> instance() {
-        return (OperatorMaterialize<T>)Holder.INSTANCE;
+        return (OperatorMaterialize<T>) Holder.INSTANCE;
     }
-    private OperatorMaterialize() { }
+
+    private OperatorMaterialize() {
+    }
+
     @Override
     public Subscriber<? super T> call(final Subscriber<? super Notification<T>> child) {
-        return new Subscriber<T>(child) {
-
+        final ParentSubscriber<T> parent = new ParentSubscriber<T>(child);
+        child.add(parent);
+        child.setProducer(new Producer() {
             @Override
-            public void onCompleted() {
-                child.onNext(Notification.<T> createOnCompleted());
-                child.onCompleted();
+            public void request(long n) {
+                if (n > 0) {
+                    parent.requestMore(n);
+                }
             }
+        });
+        return parent;
+    }
 
-            @Override
-            public void onError(Throwable e) {
-                RxJavaPlugins.getInstance().getErrorHandler().handleError(e);
-                child.onNext(Notification.<T> createOnError(e));
-                child.onCompleted();
+    private static class ParentSubscriber<T> extends Subscriber<T> {
+
+        private final Subscriber<? super Notification<T>> child;
+
+        // guarded by this
+        private Notification<T> terminalNotification;
```

Comment:
This can be made volatile and there is no need to synchronize to read its value.